### PR TITLE
refactor(core): introduce DI & IoC in auth and encryption modules + prepare for layering

### PR DESCRIPTION
**Pull request by mohamad-jaafar-monty**
# Description:
This PR introduces Dependency Injection (DI) and Inversion of Control (IoC) to the authentication and encryption utilities, and begins establishing a clean separation between service and controller layers.

- Related Jira:
  - [DAPA-46](https://montyholding.atlassian.net/browse/DAPA-46)

- Depends on:
  - https://github.com/montymobile1/devdox-ai-portal-api/pull/30

---

# Goals:
- Decouple core modules from hardcoded dependencies
- Improve testability by allowing injection of fakes/stubs
- Prepare for service-controller separation (future architectural layering)
- Establish protocols for key abstractions (e.g., `IUserAuthenticator`, `IEncryptionHelper`

---

# Key Changes:

## `app/utils/auth.py`
- Refactored to support two modes:
 - Old: tightly coupled `get_current_user` using hardcoded `authenticate_request`
 - New: `get_authenticated_user` accepting `IUserAuthenticator` via DI
- Introduced `UserClaims`, `IUserAuthenticator`, `ClerkUserAuthenticator`
- Added DI-ready factory: `get_user_authenticator_dependency`

## `tests/unit_test/app/utils/test_auth.py`
- Refactored tests to cover both old and new auth flows
- Introduced structured tests for edge cases and invalid headers

## `app/utils/encryption.py`
- Introduced `IEncryptionHelper` protocol
- Created `FernetEncryptionHelper` class (DI-ready)
- Maintained `EncryptionHelper` for backward compatibility (to be deprecated)

## `tests/unit_test/app/utils/test_encryption.py`
- Added test coverage for `FernetEncryptionHelper` (unit + edge cases)

## `tests/test_doubles/utils/auth/fake_authenticator.py`
- Added fake authenticators simulating:
 - Successful auth
 - Failure with generic exception
 - Clerk-style unauthorized error

## `tests/test_doubles/utils/auth/stub_http_header.py`
- Provides reusable header stubs for testing malformed/missing bearer schemes

## `app/repositories/git_label_repository.py`
- Introduced `IGitLabelStore` and default `TortoiseGitLabelStore` for future repo decoupling

---

# Architecture Notes:
- Both `auth` and `encryption` modules now support inversion of control
- Protocols enforce contract-based design, enabling future mocking or swapping of implementations
- This forms the foundation for future route/service/DB separation without rewriting business logic

---

# Tests Added/Updated:
- Unit tests for `get_current_user` and `get_authenticated_user`
- Unit tests for `FernetEncryptionHelper`
- Doubles and stubs for all injected dependencies
- Full suite covers failure, malformed input, exception propagation

---
# Deprecation Plan:
- `AuthenticatedUserDTO`-based logic will be deprecated in favor of `UserClaims`
- `EncryptionHelper` will be replaced by `FernetEncryptionHelper` once integration is complete

---

[DAPA-46]: https://montyholding.atlassian.net/browse/DAPA-46?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

- **New Features**
  - Introduced a centralized and structured exception handling system with custom error classes and HTTP status codes for improved API error responses.
  - Added a new, interface-based authentication system with user claim extraction and unified custom error handling.
  - Implemented new encryption helpers supporting user-specific encryption with enhanced key derivation and management.

- **Improvements**
  - Standardized API responses to consistently return JSON with enhanced serialization of Pydantic models and richer error details.
  - Centralized registration of exception handlers for streamlined error management.
  - Standardized error message constants across the application.
  - Improved logging configuration with consistent formatting and traceability.

- **Bug Fixes**
  - Fixed type hints and explicit JSON response returns in API endpoints.
  - Corrected indentation and formatting issues in multiple modules.

- **Tests**
  - Added extensive unit tests covering exception handling, authentication flows, encryption helpers, and API response serialization.
  - Introduced new test doubles, stubs, and fixtures for authentication and error scenarios.
  - Removed outdated and deprecated test cases, and updated existing tests to reflect new error handling.

- **Refactor**
  - Refactored authentication and encryption utilities for better clarity, extensibility, and testability.
  - Reorganized constants and imports for improved code structure and consistency.

- **Chores**
  - Enhanced code formatting and whitespace consistency throughout the codebase.
  - Updated test and coverage scripts for cleaner and more maintainable output.

### CODE REVIEW

File: app/middlewares/log_trace_context.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+# during the lifecycle of a single request.
+trace_id_var: ContextVar[str] = ContextVar("trace_id", default="-")
+user_id_var: ContextVar[str] = ContextVar("user_id", default="-")
+
+
+def get_trace_id() -> str:
+    return trace_id_var.get()
+
+
+def get_user_id() -> str:
```
Comment (Hayat55) [inline app/middlewares/log_trace_context.py]: this function and get_trace_id  are called for every log record and will affect the performance on production

Comment (Hayat55) [inline app/middlewares/log_trace_context.py]: The existing logging implementation introduces memory and CPU overhead. Specifically, each call to ContextVar allocates memory, leading to increased memory consumption. Additionally, frequent string formatting and context switching contribute to CPU overhead.

**Recommendation:**
Refactor the setup_logging function to accept a strategy parameter, allowing different logging behaviors depending on the environment or use case. The supported strategies could include:

"async": Use async strategy to minimize overhead — suitable for production environments.

"lazy": Enable lazy evaluation of log messages for dynamic or conditional log content.

"level": Attach context only to logs at WARNING level or higher — ideal for minimal error tracking.

"original": Retain the original implementation — useful during development for full verbosity and flexibility.

**Review by Hayat55 — CHANGES_REQUESTED** [review]

File: app/middlewares/log_trace_context.py (hunk 1 above)
Comment (mohamad-jaafar-monty) [reply app/middlewares/log_trace_context.py]: How do you propose I proceed with this. This is the current behaviour: 
- I add `trace_id` to every log. When it's empty I set it to `-`. When its not it shows a `uuid`. 
- `trace_id` is set before every request, based on the 2nd comment it seems this is not the behaviour you want.

File: app/routes/git_tokens.py [hunk 2]
```diff
@@ -262,41 +258,26 @@ async def add_git_token(
     request: Request,
     payload: GitLabelBase = Body(...),
     authenticated_user: AuthenticatedUserDTO = CurrentUser,
-) -> Dict[str, Any]:
+) -> JSONResponse | dict[str, Any]:
     """
     Add a new git token configuration with validation based on hosting service.
     """
     try:
         # Override user_id with authenticated user ID for security
```
Comment (Hayat55) [inline app/routes/git_tokens.py]: remove this comment as no need for it anymore

File: app/middlewares/log_trace_context.py (hunk 1 above)
Comment (mohamad-jaafar-monty) [reply app/middlewares/log_trace_context.py]: Generally speaking, do you want to show the `trace_id` for every request, or only when an error occurs. What is the ideal case scenario for you. Both are fine by me.

File: app/routes/git_tokens.py [hunk 3]
```diff
@@ -262,41 +258,26 @@ async def add_git_token(
...
-        # payload.user_id = authenticated_user.id
```
Comment (Hayat55) [inline app/routes/git_tokens.py:271]: same for this comment

**Review by Hayat55 — CHANGES_REQUESTED** [review]

File: app/middlewares/log_trace_context.py (hunk 1 above)
Comment (Hayat55) [reply app/middlewares/log_trace_context.py]: yes keep it please for each request

File: app/services/logging_service/logging_config.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    Notes:
+        - This affects all loggers created via `logging.getLogger()`
+        - Should be called once at app startup
+    """
+
+    root_logger = logging.getLogger()
+
+    root_logger.setLevel(getattr(logging, settings.LOG_LEVEL.upper()))
+
```
[the remark this answers is not in the document]
Comment (mohamad-jaafar-monty) [reply app/services/logging_service/logging_config.py]: It is intentional, its meant to prevent hidden configuration changes, to show that there is an issue, rather than allowing it to fester in the background.

File: app/services/logging_service/context.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+            if RequestContext._is_trace_cached:
+                trace_id = RequestContext._cache.get("trace_id", "no-trace")
+                parts.append(f"[trace_id={trace_id}]")
+
+            if RequestContext._is_user_cached:
+                user_id = RequestContext._cache.get("user_id", "unknown")
+                parts.append(f"[user_id={user_id}]")
+
+            return " ".join(parts) + " " + msg, kwargs
```
[the remark this answers is not in the document]
Comment (mohamad-jaafar-monty) [reply app/services/logging_service/context.py]: Fine, will implement it.

File: app/repositories/git_label_repository.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+from app.models import GitLabel
+
+
+class IGitLabelStore(Protocol):
+    async def get_by_id_and_user(
+        self, token_id: str, user_id: str
+    ) -> Optional[GitLabel]: ...
+
+
+class TortoiseGitLabelStore(IGitLabelStore):
```
Comment (Hayat55) [inline app/repositories/git_label_repository.py:12]: are we using it ?

File: app/utils/auth.py [hunk 7]
```diff
@@ -1,134 +1,179 @@
...
+        payload = auth_result.payload
+        user = UserClaims(**payload)
+        return user
+
+
+def get_user_authenticator_dependency() -> IUserAuthenticator:
+    return ClerkUserAuthenticator()
+
+
+async def get_authenticated_user(
```
Comment (Hayat55) [inline app/utils/auth.py:172]: when we are calling this function ?

File: tests/test_repos.py [hunk 8]
```diff
@@ -128,6 +129,7 @@ def mock_git_label():
 )
 
 
+@pytest.mark.skip("OUTDATED WILL BE REMOVED IN FUTURE")
 class TestGetReposEndpoint:
```
[the remark this answers is not in the document]
Comment (mohamad-jaafar-monty) [reply tests/test_repos.py:133]: Will be removed in future pull requests, not in this one.

File: app/repositories/git_label_repository.py (hunk 6 above)
Comment (mohamad-jaafar-monty) [reply app/repositories/git_label_repository.py:12]: we are using it in the 2nd repo get, but it isn't implemented in this PR, but in a next PR.

File: app/utils/auth.py (hunk 7 above)
Comment (mohamad-jaafar-monty) [reply app/utils/auth.py:172]: in `app/routes/repos.py` specifically in this part of the code:
```python
@router.get(
    "/",
    response_model=RepoListResponse,
    status_code=status.HTTP_200_OK,
    summary="Get all repos",
    description="Retrieve a paginated list of repositories for a user",
)
async def get_repos(
    user: UserClaims = Depends(get_authenticated_user), # <--------||-------- Called here
    ......
) -> JSONResponse:

    .......
```

File: app/repositories/git_label_repository.py (hunk 6 above)
Comment (Hayat55) [reply app/repositories/git_label_repository.py:12]: it looks like you missed commit this file
